### PR TITLE
Unified query converter for Visibility

### PR DESCRIPTION
**Pull request by rodrigozhou**
## What changed?
Unify Visibility query converters.

The new query converter is the struct `query.QueryConverter[ExprT]`. The generic parameter specify the output type of the converter (in ES is `elastic.Query`, in SQL is `sqlparser.Expr`).

The query converter has a `query.StoreQueryConverter[ExprT]`. This is the Visibility store specific implementation for building the output query.

For example, the store needs to implement `BuildAndExpr(exprs)` to tell the query converter how to build the `AND` expression. In SQL, it's `*sqlparser.AndExpr{exprs}`. In Elasticsearch, it's `elastic.NewBoolQuery().Filter(exprs)`.

The store query converter for SQL needs a plugin query converter called `sqlplugin.VisibilityQueryConverter` which is implemented by each plugin (MySQL, PostgreSQL, SQLite). This interface is needed to build DB specific syntax for `KeywordList` and `Text` searches as well as to build the final `SELECT` statements.

Added dynamic config `system.VisibilityEnableUnifiedQueryConverter` that acts as switch between the legacy and the unified query converters.

## Why?
There are two query converters: one for Elasticsearch, and another for SQL.
They are somewhat similar, and have quite a few code duplications for validation, modifying the query, etc.
Unifying the query converter ensures that we provide the exact same behavior across different Visibility stores.

## How did you test it?
- [x] built
- [x] run locally and tested manually
- [x] covered by existing tests
- [x] added new unit test(s)
- [ ] added new functional test(s)

I've added a lot of unit tests for the new code with coverage close to 100%.

Note that there has been basically no changes to functional tests, which suggests that the change works well.

## Potential risks
There might be some edge cases that either query converters that this unified query converter might deal differently.

### CODE REVIEW

File: common/persistence/visibility/store/query/interceptors.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package query
+
+type SearchAttributeInterceptor interface {
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/interceptors.go:9]: Suggest a comment on the semantics of this interceptor, and what is it used for.

File: common/persistence/visibility/store/query/nil_converter.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
+package query
+
+import (
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/searchattribute"
+)
+
+type nilStoreQueryConverter struct{}
+
+var _ StoreQueryConverter[any] = (*nilStoreQueryConverter)(nil)
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/nil_converter.go:14]: Where is this StoreQueryConverter defined?

File: common/persistence/visibility/store/query/util.go [hunk 3]
```diff
@@ -2,11 +2,133 @@ package query
...
 
+type (
+	// UnsafeSQLString don't escape the string value; unlike sqlparser.SQLVal.
+	// This is used for building string known to be safe.
+	UnsafeSQLString struct {
+		sqlparser.Expr
+		Val string
+	}
+
+	ColName struct {
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/util.go]: Should we call this as Column instead of ColName? Similarly, SAColumn instead of SAColName?
Rationale: this struct encapsulates a column and associated methods.

File: common/persistence/visibility/store/query/util.go [hunk 4]
```diff
@@ -2,11 +2,133 @@ package query
...
+		sqlparser.Expr
+		Name string
+	}
+
+	SAColName struct {
+		sqlparser.Expr
+		Alias     string
+		FieldName string
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/util.go:32]: Comment that FieldName is the actual column name in the table.

File: common/persistence/visibility/store/query/util.go [hunk 5]
```diff
@@ -22,3 +144,22 @@ func ParseExecutionDurationStr(durationStr string) (time.Duration, error) {
...
 }
+
+func ReduceExprs(
+	reduceFunc func(left, right sqlparser.Expr) sqlparser.Expr,
+	exprs ...sqlparser.Expr,
+) sqlparser.Expr {
+	if len(exprs) == 0 {
+		return nil
+	}
+	for n := len(exprs); n > 1; n = (n + 1) / 2 {
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/util.go]: Just curious as to why we prefer a pairwise reduction compared to a single left fold approach like this:
The pairwise is useful if we want to parallelize the operations, but I don't see that happening here.

```
result := exprs[0]
for i := 1; i < len(exprs); i++ {
    result = reduceFunc(result, exprs[i])
}
```

File: common/persistence/visibility/store/sql/query_converter_legacy_factory.go [hunk 6]
```diff
@@ -8,7 +8,7 @@ import (
 	"go.temporal.io/server/common/searchattribute"
 )
 
-func NewQueryConverter(
+func NewQueryConverterLegacy(
```
Comment (rkannan82) [inline common/persistence/visibility/store/sql/query_converter_legacy_factory.go:11]: Shouldn't this change and visibility_store.go (below) be part of this parent pull request?
https://github.com/temporalio/temporal/pull/8333:

File: common/persistence/visibility/store/query/nil_converter.go
Comment (rkannan82) [inline common/persistence/visibility/store/query/nil_converter.go:1]: Would it be useful to add a file comment that says:  this is a dummy converter that is used for collecting query metadata via interceptors.

File: service/worker/scheduler/query.go [hunk 7]
```diff
@@ -37,13 +38,44 @@ func newFieldNameAggInterceptor(
 	}
 }
 
+type saAggInterceptor struct {
+	names map[string]bool
```
Comment (rkannan82) [inline service/worker/scheduler/query.go]: Suggest a comment stating what this is storing (key and value).
Reason for asking is: The implementation is adding an entry to the map when a column was referenced in the query. It is not clear why we need the bool though. As in, can the value be just an empty struct?

File: common/persistence/sql/sqlplugin/interfaces.go [hunk 8]
```diff
@@ -31,6 +31,8 @@ type (
 	// Plugin defines the interface for any SQL database that needs to implement
 	Plugin interface {
 		CreateDB(dbKind DbKind, cfg *config.SQL, r resolver.ServiceResolver, l log.Logger, mh metrics.Handler) (GenericDB, error)
+
+		GetVisibilityQueryConverter() VisibilityQueryConverter
```
Comment (rkannan82) [inline common/persistence/sql/sqlplugin/interfaces.go:35]: Why do we have this Visibility prefix here?

File: common/persistence/sql/sqlplugin/postgresql/query_converter.go
Comment (rkannan82) [inline common/persistence/sql/sqlplugin/postgresql/query_converter.go:1]: Why isn't this code branched from common/persistence/visibility/store/sql/query_converter_postgresql.go.
Same for other files. I was expecting fewer diff.

File: common/persistence/visibility/store/query/util.go (hunk 5 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/util.go]: Yes, no good reason besides keeping the tree height smaller. Just personal preference, nothing against doing left fold.

File: common/persistence/visibility/store/sql/query_converter_legacy_factory.go (hunk 6 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/sql/query_converter_legacy_factory.go:11]: This PR is stacked on top of #8333, that's why that PR changes is also reflected here.

File: service/worker/scheduler/query.go (hunk 7 above)
Comment (rodrigozhou) [reply service/worker/scheduler/query.go]: Yes, sure. We don't use the bool value, this is more meant to be a set, but Go doesn't have a native set type.

File: common/persistence/sql/sqlplugin/interfaces.go (hunk 8 above)
Comment (rodrigozhou) [reply common/persistence/sql/sqlplugin/interfaces.go:35]: Just to be explicit that this query converter inside the `Plugin` is for Visibility. The `Plugin` interface is used to represent the whole DB implementation for both the core and Visibility stores.

File: common/persistence/sql/sqlplugin/postgresql/query_converter.go
Comment (rodrigozhou) [reply common/persistence/sql/sqlplugin/postgresql/query_converter.go:1]: The code is branched from those files actually, but since it's a new file, not renaming, git is showing just as new code.

File: common/persistence/visibility/store/query/nil_converter.go (hunk 2 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/nil_converter.go:14]: It's defined in `common/persistence/visibility/store/query/converter.go`

File: common/persistence/visibility/store/query/util.go (hunk 5 above)
Comment (rkannan82) [reply common/persistence/visibility/store/query/util.go]: - In this impl, what is the significant of tree height?
- The left fold impl is quite simple. That's why I suggested it in favor of this logic. You can make the final call.

File: common/persistence/visibility/store/query/util.go [hunk 9]
```diff
@@ -2,11 +2,136 @@ package query
...
 
+type (
+	// UnsafeSQLString don't escape the string value; unlike sqlparser.SQLVal.
+	// This is used for building string known to be safe.
+	UnsafeSQLString struct {
+		sqlparser.Expr
+		Val string
+	}
+
+	ColumnName struct {
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/util.go:22]: I thought I left a comment earlier on calling this as Column instead of ColumnName. Similarly, SAColumn. 
Rationale was it encapsulates more than just a name.

File: common/persistence/visibility/store/query/converter.go [hunk 10]
```diff
@@ -0,0 +1,713 @@
+//go:generate mockgen -package $GOPACKAGE -source $GOFILE -destination converter_mock.go
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/converter.go:1]: Delete this?

File: common/persistence/visibility/store/query/converter.go [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/primitives"
+	"go.temporal.io/server/common/searchattribute"
+	"go.temporal.io/server/common/sqlquery"
+)
+
+type (
+	StoreQueryConverter[ExprT any] interface {
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/converter.go:29]: Suggest adding some comments to this interface and the method. For the methods, I think it suffices to give an example of what the input is and what the output is. This will also help me verify the impl.

Sorry, I missed reviewing this file earlier because the diff was collapsed :)

File: common/persistence/visibility/store/query/converter.go [hunk 12]
```diff
@@ -0,0 +1,713 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return queryParams, nil
+}
+
+func (c *QueryConverter[ExprT]) convertWhereString(queryString string) (*QueryParams[ExprT], error) {
+	where := strings.TrimSpace(queryString)
+	if where != "" &&
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/converter.go:196]: Should we return an error if the queryString has where clause in it?

File: common/persistence/visibility/store/query/converter.go [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+
+	//nolint:revive // type cast is guaranteed to be a *sqlparser.Select
+	selectStmt, _ := stmt.(*sqlparser.Select)
+	return c.convertSelectStmt(selectStmt)
+}
+
+func (c *QueryConverter[ExprT]) convertSelectStmt(
+	sel *sqlparser.Select,
+) (*QueryParams[ExprT], error) {
+	// if sel.OrderBy != nil {
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/converter.go:218]: delete?

File: common/persistence/visibility/store/query/converter.go [hunk 14]
```diff
@@ -0,0 +1,713 @@
...
+	if sel.Where == nil {
+		sel.Where = &sqlparser.Where{
+			Type: sqlparser.WhereStr,
+			Expr: nil,
+		}
+	}
+
+	res := &QueryParams[ExprT]{}
+	if sel.Where.Expr != nil {
+		// // Wrap the WHERE clause in parenthesis to make sure the added conditions below (namespace and
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/converter.go]: delete?

File: common/persistence/visibility/store/query/converter.go [hunk 15]
```diff
@@ -0,0 +1,713 @@
...
+			"%s: 'GROUP BY' clause supports only a single field",
+			NotSupportedErrMessage,
+		)
+	}
+	for k := range sel.GroupBy {
+		colName, err := c.convertColName(sel.GroupBy[k])
+		if err != nil {
+			return nil, err
+		}
+		if colName.FieldName != searchattribute.ExecutionStatus {
```
Comment (rkannan82) [inline common/persistence/visibility/store/query/converter.go]: Why only ExecutionStatus is allowed? What about other search attributes?

File: common/persistence/sql/sqlplugin/interfaces.go (hunk 8 above)
Comment (rkannan82) [reply common/persistence/sql/sqlplugin/interfaces.go:35]: Sorry, still unclear. Are you saying for a core store, we need to add another API called GetCoreQueryConverter to this interface?

**Review by rkannan82 — APPROVED** [review]

File: service/worker/scheduler/query.go (hunk 7 above)
Comment (rkannan82) [reply service/worker/scheduler/query.go]: Could you just use an empty struct as value instead? is that an idiomatic way to represent a set?

File: common/persistence/sql/sqlplugin/postgresql/query_converter.go
Comment (rkannan82) [reply common/persistence/sql/sqlplugin/postgresql/query_converter.go:1]: Let me sync up with you offline on this. For example, could we have done this in 2 steps:
1. Copy the file. Noithing to review.
2. Make the changes? This shows just the diff?

I am just worried there is a lot of code.

File: common/persistence/sql/sqlplugin/interfaces.go (hunk 8 above)
Comment (rodrigozhou) [reply common/persistence/sql/sqlplugin/interfaces.go:35]: Discussed offline

File: common/persistence/visibility/store/query/converter.go (hunk 12 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/converter.go:196]: What do you mean?

File: common/persistence/visibility/store/query/converter.go (hunk 10 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/converter.go:1]: This is to auto-generate the `converter_mock.go` file which is used in tests.

File: common/persistence/visibility/store/query/converter.go (hunk 15 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/converter.go]: We only allow group by ExecutionStatus at this moment because of perf concerns.

File: common/persistence/visibility/store/query/util.go (hunk 9 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/util.go:22]: `ColumnName` just encapsulate a column name. I can rename `SAColumnName` to `SAColumn`.

File: common/persistence/visibility/store/query/converter.go (hunk 13 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/converter.go:218]: I added a comment that we want to uncomment these lines in a future release.

File: common/persistence/visibility/store/query/converter.go (hunk 11 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/converter.go:29]: Added comments

File: common/persistence/visibility/store/query/util.go (hunk 5 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/util.go]: Changed to left fold.

File: common/persistence/visibility/store/query/converter.go (hunk 15 above)
Comment (rkannan82) [reply common/persistence/visibility/store/query/converter.go]: Optional. Externalize the allowed attributes in a constant with a comment and use it here? Then it is obvious.

File: common/persistence/visibility/store/query/converter.go (hunk 12 above)
Comment (rkannan82) [reply common/persistence/visibility/store/query/converter.go:196]: This code assumes the string does not contain where clause. We implicitly add it inside this function. So was asking if we should validate the string does not have it.

Comment (rodrigozhou) [reply common/persistence/visibility/store/query/converter.go:196]: If it has the `where` string, the parser will fail.
